File: Difficulty: Medium/Longest Increasing Subsequence/longest-increasing-subsequence.java
class Solution {
    static int lis(int arr[]) {
        int n = arr.length;
        
        int[] prefix = new int[n];
        prefix[0] = 1;
        
        int max = prefix[0];
        for (int i = 1; i < n; i++) {
            prefix[i] = 1;
            
            int maxPrev = 0;
            for (int j = i - 1; j >= 0; j--) {
                if (arr[j] < arr[i]) {
                    if (prefix[j] > maxPrev) {
                        maxPrev = prefix[j];
                    }
                }
            }
            
            prefix[i] += maxPrev;
            
            if (prefix[i] > max) {
                max = prefix[i];
            }
        }
        
        // System.out.println(Arrays.toString((prefix)));
        
        return max;
    }
}

File: Difficulty: Medium/Longest Increasing Subsequence/longest-increasing-subsequence.py
class Solution:
    def search(self, arr, target):
        n = len(arr)
        for i in range(0, n):
            if arr[i] >= target:
                break
        
        return i
        
        
    def lis(self, arr):
        tmp = []
        
        n = len(arr)    
        for i in range(0, n):
            if len(tmp) == 0:
                tmp.append(arr[i])
                continue
            
            if arr[i] > tmp[-1]:
                tmp.append(arr[i])
                continue
            
            j = self.search(tmp, arr[i])
            tmp[j] = arr[i]
        
        return len(tmp)
       
